### PR TITLE
Allow overriding a sensor's display precision

### PR DESCRIPTION
**Pull request by emontnemery**
## Proposed change

Allow overriding a sensor's display precision

Needs core PR https://github.com/home-assistant/core/pull/87539

image

image

## Type of change

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (thank you!)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Example configuration

```yaml

```

## Additional information

- This PR fixes or closes issue: fixes #
- This PR is related to issue or discussion:
- Link to documentation pull request:

## Checklist

- [ ] The code change is tested and works locally.
- [ ] There is no commented out code in this PR.
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

[docs-repository]: https://github.com/home-assistant/home-assistant.io

### CODE REVIEW

File: src/common/number/format_number.ts [hunk 1]
```diff
@@ -90,8 +91,15 @@ export const formatNumber = (
  * @returns An `Intl.NumberFormatOptions` object with `maximumFractionDigits` set to 0, or `undefined`
  */
 export const getNumberFormatOptions = (
-  entityState: HassEntity
+  entityState: HassEntity,
+  entity: EntityRegistryEntry | undefined
 ): Intl.NumberFormatOptions | undefined => {
+  const display_precision =
+    entity?.options?.sensor?.display_precision ||
+    entity?.options?.sensor?.suggested_display_precision;
```
Comment (piitaya) [inline src/common/number/format_number.ts:99]:
```suggestion
    entity?.options?.sensor?.display_precision ??
    entity?.options?.sensor?.suggested_display_precision;
```
We must check for `null` and `undefined` not falsy.

File: src/common/number/format_number.ts [hunk 2]
```diff
@@ -90,8 +91,15 @@ export const formatNumber = (
  * @returns An `Intl.NumberFormatOptions` object with `maximumFractionDigits` set to 0, or `undefined`
  */
 export const getNumberFormatOptions = (
-  entityState: HassEntity
+  entityState: HassEntity,
+  entity: EntityRegistryEntry | undefined
```
Comment (piitaya) [inline src/common/number/format_number.ts:95]:
```suggestion
  entityState: HassEntity,
  entity?: EntityRegistryEntry
```
So you can avoid passing `undefined` in tests.

File: src/panels/config/entities/entity-registry-settings.ts [hunk 3]
```diff
@@ -277,6 +287,14 @@ export class EntityRegistrySettings extends SubscribeMixin(LitElement) {
     }
   }
 
+  private precisionLabel(precision: number, _state?: string) {
+    const value = _state ?? 0;
```
Comment (bramkragten) [inline src/panels/config/entities/entity-registry-settings.ts:291]:
```suggestion
  private precisionLabel(precision: number, state?: string) {
    const value = state ?? 0;
```

File: src/panels/config/entities/entity-registry-settings.ts [hunk 4]
```diff
@@ -468,6 +486,44 @@ export class EntityRegistrySettings extends SubscribeMixin(LitElement) {
...
+                )}
+                .value=${this._precision == null
+                  ? "default"
+                  : this._precision.toString()}
+                naturalMenuWidth
+                fixedMenuPosition
+                @selected=${this._precisionChanged}
+                @closed=${stopPropagation}
+              >
+                <mwc-list-item .value=${"default"}
```
Comment (bramkragten) [inline src/panels/config/entities/entity-registry-settings.ts:509]: use `ha-list-item`

File: src/panels/lovelace/common/has-changed.ts [hunk 5]
```diff
@@ -24,6 +26,37 @@ function hasConfigChanged(element: any, changedProps: PropertyValues): boolean {
   return false;
 }
 
+function compareEntityState(
+  oldHass: HomeAssistant,
+  newHass: HomeAssistant,
+  entityId: string
+) {
+  const oldState = oldHass.states[entityId] as HassEntity | undefined;
+  const newState = newHass!.states[entityId] as HassEntity | undefined;
```
Comment (bramkragten) [inline src/panels/lovelace/common/has-changed.ts:35]:
```suggestion
  const newState = newHass.states[entityId] as HassEntity | undefined;
```

File: src/panels/lovelace/common/has-changed.ts [hunk 6]
```diff
@@ -51,12 +85,18 @@ export function hasConfigOrEntitiesChanged(
...
-      oldHass.states[entity.entity] !== element.hass!.states[entity.entity]
-  );
+  return entities.some((entity) => {
+    if (!("entity" in entity)) {
+      return false;
+    }
+
+    return (
+      compareEntityState(oldHass, newHass, element._config!.entity) ||
+      compareEntityEntryOptions(oldHass, newHass, element._config!.entity)
```
Comment (bramkragten) [inline src/panels/lovelace/common/has-changed.ts:99]:
```suggestion
      compareEntityState(oldHass, newHass, entity.entity) ||
      compareEntityEntryOptions(oldHass, newHass, entity.entity)
```

**Review by bramkragten — APPROVED** [review]

**Comment by bramkragten** [comment]
We should really split up `entity-registry-settings.ts` in the future per domain, it is getting big and bulky 😬

**Comment by piitaya** [comment]
> We should really split up `entity-registry-settings.ts` in the future per domain, it is getting big and bulky 😬

I will create a clickup for that

**Pull request merged**